controller/persistence/eventual_eventer.go

```diff
@@ -0,0 +1,607 @@
+	var ids []string
+	var eventualEvents []*EventualEvent
+	err := a.dbProvider.GetDb().View(func(tx *bbolt.Tx) error {
+		var err error
+		ids, _, err = a.store.QueryIds(tx, fmt.Sprintf("limit %d", a.batchSize))
+
+		if err != nil {
+			return err
+		}
+
+		for _, id := range ids {
+			event, err := a.store.LoadOneById(tx, id)
+
+			if err != nil {
+				pfxlog.Logger().WithField("id", id).WithError(err).Errorf("error could not load event id %s", id)
+			} else {
+				eventualEvents = append(eventualEvents, event)
+			}
+		}
+		return nil
+	})
+
+	return ids, eventualEvents, err
+}
+
+// process is the main execution look for dealing with batches of eventual events, triggering listeners for
+// the eventual events and emitting normal processing events
+func (a *EventualEventerBbolt) process() {
+	processId := cuid.New()
+	totalBatches := int64(0)
+	totalEvents := int64(0)
+	totalListenersExecuted := int64(0)
+	totalEventIndex := int64(0)
+
+	a.Emit(EventualEventProcessingStartName, &EventualEventProcessingStart{
+		Id: processId,
+	})
+
+	defer func() {
+		a.notifyWaiters()
+
+		a.Emit(EventualEventProcessingDoneName, &EventualEventProcessingDone{
+			Id:                     processId,
+			TotalBatches:           totalBatches,
+			TotalEvents:            totalEvents,
+			TotalListenersExecuted: totalListenersExecuted,
+		})
+	}()
+
+	for {
+		batchId := cuid.New()
+
+		var numIds = 0
+		var ids []string
+		var eventualEvents []*EventualEvent
+
+		// if we know we have outstanding events query for them up to 5 times then give up
+		numEvents := atomic.LoadInt64(a.outstandingEvents)
+		for attempt := 0; attempt < 5 && numEvents > 0; attempt++ {
+			var err error
+			ids, eventualEvents, err = a.getEventualEvents()
+			numIds = len(ids)
+
+			if err != nil {
+				pfxlog.Logger().WithError(err).Error("error during EventualEventerBbolt processing, could not query events")
+			}
+
+			if numIds > 0 {
+				break
+			}
+
+			time.Sleep(500 * time.Microsecond)
+			numEvents = atomic.LoadInt64(a.outstandingEvents)
+		}
+
+		if numIds == 0 {
+			return
+		}
+
+		totalBatches++
+		a.Emit(EventualEventProcessingBatchStartName, &EventualEventProcessingBatchStart{
+			Id:        batchId,
+			ProcessId: processId,
+			Count:     numIds,
+			BatchSize: a.batchSize,
+		})
+
+		if numIds != len(eventualEvents) {
+			pfxlog.Logger().
+				WithFields(map[string]interface{}{
+					"queriedCount": numIds,
+					"eventCount":   len(eventualEvents),
+				}).
+				Warnf("%d event records were queried and only %d events were loaded", numIds, len(eventualEvents))
+		}
+
+		batchEventIndex := int64(0)
+		for _, eventualEvent := range eventualEvents {
+			batchEventIndex++
+			totalEventIndex++
+
+			if handlerVals, found := a.handlerMap.Get(eventualEvent.Type); found {
+				handlers := handlerVals.([]EventListenerFunc)
+				for _, handler := range handlers {
+					func() {
+						listenerExecId := cuid.New()
```

Comment:
why a lambda and not just some function to call?

Reply:
defer'ed panic handler - so that an event handler can't kill the eventual eventer.